s3access/select_spec.py

```diff
@@ -0,0 +1,232 @@
+
+    def params(self):
+        kwargs = {}
+        if self.type is not None:
+            kwargs['Type'] = self.type.value
+        return {InputSerialization.JSON.value: kwargs}
+
+
+DEFAULT_JSON_INPUT = JsonInput()
+
+
+AnyInput = Union[ParquetInput, CsvInput, JsonInput]
+
+
+class CompressionType(Enum):
+    NONE = 'NONE'
+    GZIP = 'GZIP'
+    BZIP = 'BZIP'
+
+
+@dataclasses.dataclass(frozen=True)
+class Input:
+    compression_type: CompressionType = CompressionType.NONE
+    input: AnyInput = DEFAULT_PARQUET_INPUT
+
+    def params(self):
+        res = self.input.params()
+        res['CompressionType'] = self.compression_type.value
+        return res
+
+
+DEFAULT_INPUT = Input()
+
+
+@dataclasses.dataclass(frozen=True)
+class CsvOutput:
+    quote_fields: QuoteField = QuoteField.ALWAYS
+    record_delimited: str = '\n'
+    field_delimiter: str = ','
+    quote_character: str = '"'
+    quote_escape_character = '"'
+
+    def params(self):
+        return {
+            OutputSerialization.CSV.value: {
+                'QuoteFields': self.quote_fields.value,
+                'RecordDelimiter': self.record_delimited,
+                'FieldDelimiter': self.field_delimiter,
+                'QuoteCharacter': self.quote_character,
+                'QuoteEscapeCharacter': self.quote_escape_character
+            }
+        }
+
+
+CSV_DEFAULT_OUTPUT = CsvOutput()
+
+
+@dataclasses.dataclass(frozen=True)
+class JsonOutput:
+    record_delimiter: str = '\n'
+
+    def params(self):
+        return {OutputSerialization.JSON.value: {'RecordDelimiter': self.record_delimiter}}
+
+
+JSON_DEFAULT_OUTPUT = JsonOutput()
+
+AnyOutput = Union[CsvOutput, JsonOutput]
+
+
+@dataclasses.dataclass(frozen=True)
+class Output:
+    output: AnyOutput = CSV_DEFAULT_OUTPUT
+
+    def params(self):
+        return self.output.params()
+
+
+DEFAULT_OUTPUT = Output()
+
+
+@dataclasses.dataclass(frozen=True)
+class ScanRange:
+    start: int = 0
+    end: Optional[int] = None
+
+    def params(self) -> Dict[str, int]:
+        res = {'Start': self.start}
+        if self.end is not None:
+            res['End'] = self.end
+        return res
+
+
+DEFAULT_SCAN_RANGE = ScanRange()
+
+
+def _quote(v) -> str:
+    if isinstance(v, str):
+        v = v.replace("'", "''")
+        return f"'{v}'"
+    if isinstance(v, (date, datetime)):
+        return f"'{v.isoformat()}'"
+    if isinstance(v, Sequence):
+        return f"({','.join(map(_quote, v))})"
+    return str(v)
+
+
+def simple_parquet_selection(
+        columns: Optional[Sequence[str]] = None,
+        filters: Sequence[Tuple[Any, str, Any]] = None) -> str:
+    """
+    select the given columns and filter the data by using "AND" between all given filters
+
+    example usage:
+    ---------------------------------------------
+    columns = ['col1', 'col2', 'col5', 'col10']  # some columns all files share
+    filters = [('col1', '=', 'some string value'), ('col5', '>=', 10.2), ('col5', '<', 20.3)]
```

Comment:
How would one do `IN`?

Reply:
That's the easiest one to answer so ...
This is modeled basically (in a much more simplistic way) to the way filtering is done at the pyarrow level ([DNF filters](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html)).
So for IN operator: 
```python
('col1', 'IN', ['a', 'b', 'b'])
```

Reply:
✅